### ISSUE TITLE
Port DataTable.AsDataView extension methods

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/corefx#27610]
These were in System.Data.DatasetExtensions but we did not port them because they need internal access to System.Data.Common (was System.Data)

Would probably be best to merge S.D.DSE into S.D.C with type forwards.

Usage seems very low but at least one porting customer needs it. See https://github.com/dotnet/corefx/issues/19771

**Comment by ktarbet** [comment]
I'm also needing DataTableExtensions in .net core for my projects.  I have copied files from link below , and hacked some String Resources named  'SR' to compile.  There may be a better way?

https://github.com/dotnet/corefx/tree/master/src/System.Data.DataSetExtensions/src

**Comment by ktarbet** [comment]
Thanks @danmosemsft  -- I was able to use the Compat pack.

**Comment by danmosemsft** [comment]
@ktarbet great! Please do open issues here with any bugs or feedback on .NET Core.

**Comment by HumanAftrAll** [comment]
Oh, boy. I need AsDataView()... Please port it all

**Comment by danmosemsft** [comment]
OK, got data from some Winforms apps where AsDataView is missing also. Tracking on the Winforms project now. Note as above it requires a typeforward to avoid the InternalsVisibleToAttribute.

**Comment by Anipik** [comment]
@ericstj @weshaggard here we have to move implementation of all the types from S.D.DSE to S.D.C and just type forwards in S.D.DSE and port t DataTable.AsDataView extension method into S.D.C . is that all or we have to do something else too ?

**Comment by ericstj** [comment]
Well, currently `DataSetExtensions` builds as netstandard and `System.Data.Common` was absorbed into netstandard.dll, so you can't really change `System.Data.Common` and use that from the netstandard library.  Moreover, the specific type to which we need internals access is DataView, to override internal virtuals, and since that type is part of netstandard we cannot build a netstandard impl of DataTableExtensions.  You'd have to build a netcoreapp specific version if you wanted to change System.Data.Common.  In other words you can only expose DataTableExtensions to netcoreapp (and you'll have to start cross-compiling).

This is all assuming the current state of the code.  It may be possible to implement this on top of the NETStandard API by changing the implementation, but that's a different level of work.  @JeremyKuhne might better understand the feasibility based on his [comment](https://github.com/dotnet/corefx/issues/19771#issuecomment-368998179).

If its not possible then we should probably look at either exposing some of those virtuals or absorbing DataTableExtensions into netstandard-vnext if we want to make the API accessible by netstandard libraries.

**Comment by JeremyKuhne** [comment]
The overriding of [internal virtuals](https://referencesource.microsoft.com/#System.Data.DataSetExtensions/System/Data/LinqDataView.cs,122) is what really made this impossible. LinqDataView is changing the way the base class does its searching.

I can't recall coming up with any solutions, even complicated ones. I think you're stuck with

> we should probably look at either exposing some of those virtuals or absorbing DataTableExtensions into netstandard-vnext if we want to make the API accessible by netstandard libraries.

**Comment by danmosemsft** [comment]
@Anipik what are the specific virtuals that would need to be exposed?

**Comment by Anipik** [comment]
The ones we require are
```c#
internal virtual int FindByKey(object key);
internal virtual int FindByKey(object[] key);
internal virtual DataRowView[] FindRowsByKey(object[] key);
internal virtual void SetIndex(string newSort, DataViewRowState newRowStates, IFilter newRowFilter);
```

we will alse need to expose this 

```c#
 internal DataView(DataTable table, System.Predicate<DataRow> predicate, System.Comparison<DataRow> comparison, DataViewRowState RowState);
```

**Comment by Anipik** [comment]
As Discussed offline  it's too difficult to do it in a good enough way, so putting it on a hold for sometime

**Comment by Anipik** [comment]
For future references, this https://github.com/dotnet/corefx/pull/32302 contains most of the work done to move SDSDE into common

**Comment by danmosemsft** [comment]
Closing. We can reopen if sufficient motivation emerges.

**Comment by bmperdue** [comment]
Can I add some motivation? Would really like to see the AsDataView added to.net Core.

**Comment by don4of4** [comment]
@bmperdue https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Data.DataSetExtensions/4.5.0-preview1-25927-02

_Install this nuget package (preview) and you will probably have what you need._

**Comment by ericstj** [comment]
Fixed in #36528
